services/cd-service/pkg/mapper/environments_config.go

```diff
@@ -152,22 +152,55 @@ func DeriveGroupName(env config.EnvironmentConfig, envName string) string {
 }
 
 func calculateEnvironmentPriorities(environments []*api.Environment) {
-	// first find the maximum:
-	var maxDistance uint32 = 0
-	for i := 0; i < len(environments); i++ {
-		maxDistance = max(maxDistance, environments[i].DistanceToUpstream)
-	}
-	// now we can assign each environment a priority
+	type EnvsByName map[string]*api.Environment
+	type Childs []string
+	type ChildsByName map[string]Childs
+	var envsByName = make(EnvsByName)
+	var childsByName = make(ChildsByName)
+	// latest is UPSTREAM, so mark them as such, and the rest as OTHER for now
+	// oherwise append us to the list of the childs of the upstream env
 	for i := 0; i < len(environments); i++ {
 		var env = environments[i]
-		if env.DistanceToUpstream == maxDistance {
-			env.Priority = api.Priority_PROD
-		} else if env.DistanceToUpstream == maxDistance-1 {
-			env.Priority = api.Priority_PRE_PROD
-		} else if env.DistanceToUpstream == 0 {
+		envsByName[env.Name] = env
+		if env.Config.Upstream.GetLatest() {
 			env.Priority = api.Priority_UPSTREAM
 		} else {
 			env.Priority = api.Priority_OTHER
+			// should we barf about a config that neither has is upstream nor has one?
```

Comment:
> // should we barf about a config that neither has is upstream nor has one?

No, we should not, at least not here.
It is an usual configuration, so we could consider adding this sort of check into `GetEnvironmentConfigsAndValidate`, which is only called on startup and stops the process in case of an error.